### DISCUSSION TITLE
Extensions for verification_uri of Device Authorization Request

### DISCUSSION THREAD

**Discussion by cgeorgilakis** [keycloak/keycloak#15004] [Ideas]
[ Device Authorization Response documentation](https://www.rfc-editor.org/rfc/rfc8628.html#section-3.2) is written :

> verification_uri
      REQUIRED.  The end-user verification URI on the authorization
      server.  The URI should be short and easy to remember as end users
      will be asked to manually type it into their user agent.

Microsoft for example does something like this with {{[https://microsoft.com/devicelogin ](https://microsoft.com/devicelogin%C2%A0) }}(forwards to https://login.microsoftonline.com/common/oauth2/deviceauth).

We are thinking for being able to configure optionally per realm a shorten url for showing in verification_uri. Redirection to Keycloak  verification URI ({server_url}/realms/{realmname}/device) would be done outside Keycloak - fe in a proxy. Are you interested? Do you have another idea for shorten verification_uri?

Moreover, [Non-Textual Verification URI Optimization
](https://www.rfc-editor.org/rfc/rfc8628.html#section-3.3.1) mention:

> When "verification_uri_complete" is included in the authorization
   response ([Section 3.2](https://www.rfc-editor.org/rfc/rfc8628.html#section-3.2)), clients MAY present this URI in a non-textual
   manner using any method that results in the browser being opened with
   the URI, such as with QR (Quick Response) codes or NFC (Near Field
   Communication), to save the user from typing the URI.
 For usability reasons, it is RECOMMENDED for clients to still display
   the textual verification URI ("verification_uri") for users who are
   not able to use such a shortcut.  Clients MUST still display the
   "user_code", as the authorization server will require the user to
   confirm it to disambiguate devices or as remote phishing mitigation
   (see [Section 5.4](https://www.rfc-editor.org/rfc/rfc8628.html#section-5.4)).

QR codes are widely used lately. Are you interested in adding QR code? I I am not wrong Keycloak uses a Java QR library already.
This could be configuration in client ( by default verification_uri as url). QR code will be return as Byte Array and client implementation would be responsible for showing it appropriately. What is your opinion?

**Comment by jbman** [comment]
The client application could also shorten the URI on their own by replacing {server_url}/realms/{realmname}/device with {shortener_service_url/device. 

Regarding the QR code idea: isn't a QR code meant to be created by the client on their own? Do you suggest Keycloak should have a resource which provides a QR code for "verification_uri_complete", so that the client depends on this service instead of a library?

**Reply by cgeorgilakis** [reply]
How can client redirect {server_url}/realms/{realmname}/device? It is a Keycloak web service returned from Keycloak web service and need to copy paste on browser.

**Reply by jbman** [reply]
The device gets an authorization response with the URI to be displayed. (see [RFC8628, section-3.2](https://www.rfc-editor.org/rfc/rfc8628.html#section-3.2)). If the device knows how to shorten the URI it can do it and display the shortened URI instead. The user enters this shortened URI in their browser and it will be redirected to the original URI which was in the Keycloak response. User will enter the user_code on this page.

Only drawback is, that the client can't just display the URI from the authorization response. It needs to shorten it before it is displayed.

**Reply by cgeorgilakis** [reply]
We have over 150 clients (most of them using device authorization) in production. It is crazy to do this in each client insted of doing it centralized in keycloak configuration. Moreover, with this no need to change json response.

**Reply by jbman** [reply]
Thank you, I see that providing a shortened URI directly in the response can improve usability without changing client software.

You would need a realm level setting "Device Authorization Verification URI" which is used instead of the Verification URI in the response. Using this feature then requires a redirect service running at the configured URL.

**Comment by cgeorgilakis** [comment]
I have created a [PR](https://github.com/keycloak/keycloak-ui/pull/4081) for this
